### ISSUE TITLE
tutorial.ipynb must be tested and roughly functional!!!

### ISSUE THREAD

**Issue by lneuhaus** [lneuhaus/pyrpl#291]

**Comment by SamuelDeleglise** [comment]
I just started looking into that.  

The current version of the tutorial is focusing on the Redpitaya class. I wonder if we are not diverting the user's attention from the Pyrpl class that he is encouraged to use instead ?

**Comment by SamuelDeleglise** [comment]
In particular, there would be a little work involved if we want to bring the Redpitaya class alone up-and-running (some problems right now with the function returning the list of inputs of dsp modules that looks for the root pyrpl instance).

Since all unittests are performed with the full pyrpl instance, I would advocate to use this within the tutorial.... 

--> I am doing that right now.

**Comment by lneuhaus** [comment]
agreed. also, i suggest that all installation stuff is moved to the documentation such that the tutorial can focus more on doing actual stuff. also, the tutorial could be separated into several parts by now.

**Comment by SamuelDeleglise** [comment]
OK, so I should probably touch at the same time the documentation and the tutorial. I think I just succesfully ran the make html for the autodoc. For the moment, do you have anything that is not autogenerated in the documentation ?

**Comment by lneuhaus** [comment]
documentation is currently distributed over the wiki, the tutorial (former installation section), the readme.md, and the autodoc files.

**Comment by SamuelDeleglise** [comment]
OK, I guess we want the autodoc to look a bit more organized right ?

**Comment by lneuhaus** [comment]
The steps towards making the actual documentation are
- check https://pythonhosted.org/an_example_pypi_project/sphinx.html and pick the most reasonable docstring format 
- describe that format on this wiki page: https://github.com/lneuhaus/pyrpl/wiki/Code:-Coding-style-guide#docstrings

- for the automatic generation, we could either make a branch 'gh-pages', or just host them in the 'doc' folder
- implement automatic autodoc generation, and describe what is done here: https://github.com/lneuhaus/pyrpl/wiki/Packaging:-AutoDoc

**Comment by lneuhaus** [comment]
see also #170

**Comment by lneuhaus** [comment]
@SamuelDeleglise 
in principle the choice between google-style () and numpy-style (http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy) is a matter of taste, even though the google-style is "recommended by Khan's academy", whatever that means. I would go with google-style because i find it easier to write (and I believe pycharm can be customized), but i would also be okay with numpy-style if you see any advantages there

**Comment by lneuhaus** [comment]
Concerning the checklist item "modify the source files for autodoc to get a reasonable organization of the documentation":
This is the main decision to make now. I believe the structure of the documentation can be inspired by other pages, such as this one: https://docs.scipy.org/doc/numpy-1.13.0/user/
I believe the other points can be postponed after that..

**Comment by SamuelDeleglise** [comment]
Sorry, I was drifting in the various links of your posts...

I am fine with Google style, as long as it is properly formatted by the sphynx extension

To start the discussion, the structure could be (I am not sure where the autodoc goes by the way...):

* Hardware installation
* Installing Pyrpl
* Quickstart Tutorial
* Basics of pyrpl --> Where we explain modules and general program organization
* Contributing  --> Where we put all the stuff on unittests and git pull requests that are currently in the tutorial

**Comment by SamuelDeleglise** [comment]
OK, they seem to have a User's guide and a Reference... The user's guide is basically what I sketched above and the 'Reference' is basically the autogenerated doc.

Also, I see that they have actually put the Developer's guide in a 3rd section. Actually the root of their sphynx doc is there;

https://docs.scipy.org/doc/numpy-1.13.0/index.html

**Comment by SamuelDeleglise** [comment]
So, we could probably have:

* User's guide 
  * Hardware installation
  * Installing Pyrpl
  * Quickstart Tutorial
  * Basics of pyrpl
* Reference guide
  * A basic hand-written page that lists the different python modules with a link to their autogenerated doc
* Developer's guide
  * Compililng the FPGA code
  * Unittests
  * Code guidelines
  * Workflow to submit code changes

**Comment by SamuelDeleglise** [comment]
Are you working on it or should I start ?

**Comment by lneuhaus** [comment]
go ahead if you like

**Comment by lneuhaus** [comment]
just pull my latest commit with the sphinx configuration

**Comment by lneuhaus** [comment]
are you workking on it now?

**Comment by lneuhaus** [comment]
otherwise ill go ahead and test the structure

**Comment by SamuelDeleglise** [comment]
I was waiting for your answer and trying to implement the validate_and_normalize unit_test in the mean time. Just let me know when I can start the doc

**Comment by lneuhaus** [comment]
ok go ahead with the unit test, ill try to generate the doc structure for now

**Comment by lneuhaus** [comment]
i have to go no, continue if you like on the last commit

**Comment by SamuelDeleglise** [comment]
Alright, you haven't created the structure yet right ?

**Comment by lneuhaus** [comment]
its there, dont you see it?

**Comment by lneuhaus** [comment]
in doc/sphinx/source/index and references

**Comment by lneuhaus** [comment]
go to doc/sphinx, type 'make html' and check the build directory

**Comment by SamuelDeleglise** [comment]
Did you push it to develop-0.9.3 ?

**Comment by SamuelDeleglise** [comment]
Got it thanks

**Comment by SamuelDeleglise** [comment]
I believe the right solution is that you eventually create an account there:
http://dont-be-afraid-to-commit.readthedocs.io/en/latest/documentation.html
it will compile automatically the documentation for us (and of course, host the website).

We also need to find a solution to store the redpitaya image files and/or list of files. I am not sure if this is supported by readthedocs

**Comment by lneuhaus** [comment]
There is a largr file storage on github we can use fir that..

**Comment by lneuhaus** [comment]
pyrpl.readthedocs.io will host the docs

however, im already running into trouble with that website to correctly get pyrpl imported because of the pyqt dependency...

**Comment by lneuhaus** [comment]
readthedocs is working and the content of the documentation is slowly improving. 

Still, our website is not very welcoming at the moment. I suggest we try to get our documentation to look more similar to https://automl.github.io/auto-sklearn/stable/

**Comment by lneuhaus** [comment]
im leaving this open until we have 

- [ ] adapted the nicer style and more welcoming startup page from https://automl.github.io/auto-sklearn/stable/ or http://docs.obspy.org/ 
- [ ] finished with the tutorial.ipynb file (now at a different location)
- the rest of documentation is covered by #170

**Comment by lneuhaus** [comment]
I think that is the right location. I think in the meantime there was the idea to convert the tutorial into the module documentation on the website, so i had regarded the tutorial as deprecated. But I agree that it is best to have both, especially if testing and the generation of the displayed notebook output can be automatically done.

**Comment by lneuhaus** [comment]
Do you want the tutorial only for yesting or for new users as well?

**Comment by SamuelDeleglise** [comment]
My idea was to test the tutorial that new users will have at their disposal such that no errors are introduced in there during development...
